### PR TITLE
77 update work

### PR DESCRIPTION
**Pull request by abdulrahman-2020**

### CODE REVIEW

File: server/controllers/work/editWork.js [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+    const { id } = await paramsValidation.validateAsync(req.params);
+    const { providerID } = req;
+    const work = await Work.findByPk(id);
+    if (!work) {
+      throw customError('work does not exist', 400);
+    }
+    if (work.userId !== +providerID) {
+      throw customError('Unauthorized', 401);
+    }
+    console.log(req.body);
```
Comment (devstm) [inline server/controllers/work/editWork.js:16]: remove console.log()

**Review by devstm — CHANGES_REQUESTED** [review]

**Review by RandSohail — APPROVED** [review]

File: server/routes/index.js [hunk 2]
```diff
@@ -1,14 +1,8 @@
 const router = require('express').Router();
 const {
-  getHomeData,
-  getWorkPerPage,
-  getProfileInfo,
-  addReview,
+  getProfileInfo, addReview, getSearchResult, getHomeData, getWorkPerPage, subServices, checkAuth,
```
Comment (devstm) [inline server/routes/index.js:3]: fix ESlent

File: server/routes/index.js [hunk 3]
```diff
@@ -21,11 +15,12 @@ router.get('/work/:providerId', getWorkPerPage);
 router.get('/subservices/:mainServiceId', subServices);
 router.post('/login', loginHandler);
 router.post('/reviews', addReview);
+router.patch('/work/editWork/', checkAuth, editWork);
```
Comment (LinaYahya) [inline server/routes/index.js:18]:
```suggestion
router.patch('/work/', checkAuth, editWork);
```

File: server/database/seeders/works.js [hunk 4]
```diff
@@ -57,7 +57,7 @@ const works = [
     title: 'تبليط ساحة مسجد',
     content: 'بالزخام والجرانيت تم الانتهاء من تبليط ساحة مسجد بعد عمل شاق',
     userId: 2,
-    image: 'https://herfa.s3.eu-west-2.amazonaws.com/2/f7454dc0-cf6a-11ec-a5f1-61119f11f019',
+    image: 'dfghjkl;',
```
Comment (LinaYahya) [inline server/database/seeders/works.js:60]: add a valid link

File: server/controllers/work/editWork.js [hunk 5]
```diff
@@ -0,0 +1,39 @@
+const { Work } = require('../../database/models');
+const { customError } = require('../errors');
+const { workEditValidation } = require('../../utils/validation');
+const { uploadImage, deleteImage } = require('../../utils');
+
+const editWork = async (req, res, next) => {
+  try {
+    const { id } = await workEditValidation.validateAsync(req.body);
+    const { title, content } = req.body;
```
Comment (LinaYahya) [inline server/controllers/work/editWork.js:9]:
```suggestion
    const { id, title, content  } = await workEditValidation.validateAsync(req.body);
```

File: server/controllers/work/editWork.js [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+    if (req.files) {
+      const { image } = req.files;
+      imgUrl = await uploadImage(image, work.userId);
+      deleteImage(work.image);
+    }
+
+    await work.update({
+      title,
+      content,
+      image: imgUrl.Location,
```
Comment (LinaYahya) [inline server/controllers/work/editWork.js:28]: if not req.files, then imgUrl will be an empty string, and `imgUrl.Location` will throw an error,
And you need to set the image value if it's not send with req.files from the request body

File: server/controllers/work/editWork.js [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+      throw customError('work does not exist', 400);
+    }
+    if (work.userId !== +providerID) {
+      throw customError('Unauthorized', 401);
+    }
+    let imgUrl = '';
+    if (req.files) {
+      const { image } = req.files;
+      imgUrl = await uploadImage(image, work.userId);
+      deleteImage(work.image);
```
Comment (LinaYahya) [inline server/controllers/work/editWork.js:22]: move this after sending the response, at the end of the controller so we don't delay user from getting a response

Comment (LinaYahya) [inline server/controllers/work/editWork.js:22]: Also the delete image take the key, which is part of the url, But not the whole url
So when the url is 
`https://herfa.s3.eu-west-2.amazonaws.com/2/f7454dc0-cf6a-11ec-a5f1-61119f11f019.png`
the key woukd be 
`2/f7454dc0-cf6a-11ec-a5f1-61119f11f019.png`
So try to split after `url.com`

File: server/test/work.test.js [hunk 8]
```diff
@@ -27,31 +27,41 @@ describe('Routes Tests GET /api/v1/work/:providerId', () => {
...
-        if (err) done(err);
-        expect(res.statusCode).toBe(401);
-        done();
-      });
-  });
+  // test('Testing for update work /api/v1/work/1 and get status 401 UnAuthorized', (done) => {
+  //   supertest(app)
+  //     .patch('/api/v1/work/1')
+  //     .set('Cookie', [`token= ${process.env.TOKEN}`])
+  //     .expect(401)
```
Comment (LinaYahya) [inline server/test/work.test.js:46]: add tests to test unauthorized user, and work not found

**Review by LinaYahya — CHANGES_REQUESTED** [review]

File: server/controllers/work/editWork.js [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+  try {
+    const {
+      id, title, content,
+    } = await workEditValidation.validateAsync({ ...req.body, workImg: req.files.workImg });
+    const { providerID } = req;
+    const work = await Work.findByPk(id);
+    if (!work) {
+      throw customError('work does not exist', 400);
+    }
+    if (work.userId !== +providerID) {
```
Comment (LinaYahya) [inline server/controllers/work/editWork.js:16]:
```suggestion
    const { userId, image, id } = work;
     if (work.userId !== +providerID) {
```
Work really updated after update, so we have to save the image value we want to get the key from, On this place

**Review by LinaYahya — CHANGES_REQUESTED** [review]

**Review by LinaYahya — APPROVED** [review]

File: server/controllers/work/editWork.js [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+const { workEditValidation } = require('../../utils/validation');
+const { uploadImage, deleteImage } = require('../../utils');
+
+const editWork = async (req, res, next) => {
+  try {
+    const {
+      id, title, content,
+    } = await workEditValidation.validateAsync({ ...req.body, workImg: req.files.workImg });
+    const { providerID } = req;
+    const work = await Work.findByPk(id);
```
Comment (devstm) [inline server/controllers/work/editWork.js:12]: why we do that this is additional request to data base which we can avoid it 
so we can read { userId, image } from request body then check 
but how can handle work doesn't exist?
to know: sequelize update method retern array has 0 if failed and one if success after ubdata we can check if zero then response user with un auth message

**Review by devstm — CHANGES_REQUESTED** [review]

**Review by devstm — APPROVED** [review]

**Pull request merged**